Add headers and "Show more templates" button to template dropdown

momo3404 opened the pull request:
Fixes #1196.

Changes proposed in this PR:
- Add the headers "Organisational Templates", "Alliance General Templates", and "Other Alliance Templates" to the template dropdown to allow for better organization and readability of the templates in the dropdown.
- Add the "Show more templates" button to the dropdown that, along with shortening the initial size of the menu to allow for better readability, can also be clicked to show "Other Alliance Templates."
- Change the template selection's dropdown type from a `select_tag` to a bootstrap `dropdown` in order to allow for the "Show more templates" button to be clicked without closing the dropdown.
- Add custom CSS and javascript for the template dropdown in order to standardize the dropdown's interface and behaviour for users of all systems and browsers.
- Refactor `templation_options_controller` (previously edited in [this PR](https://github.com/portagenetwork/roadmap/pull/1199)) to simplify template handling by returning a structured JSON payload.
- Edit `plans_controller` to restrict templates that can be associated with newly created plans. Now, only templates that are published and unarchived are allowed.

[Screencast from 2025-11-12 11-23-36.webm](https://github.com/user-attachments/assets/374f11e2-0a90-43cf-875b-6aaf1cc09bf7)

momo3404 committed:
Replace select_tag in template dropdown with bootstrap dropdown in html

- select_tag has an issue where a dropdown cannot be kept open after a button is clicked. In order for the "show more templates" button to eventually work intuitively, this behaviour has to be changed.
- Bootstrap dropdowns can be kept open after a certain selection, and so that was the chosen replacement.

momo3404 committed:
Update new.js to work with bootstrap dropdown behaviour

- Edit toggleSubmit() so that both select tags and bootstrap dropdowns can still work with the function. The previous behaviour only worked with select tags.
- Edit the success() function to work with the bootstrap dropdown.

momo3404 committed:
Add CSS for bootstrap template dropdown

momo3404 committed:
Add DEFAULT_FUNDER_ID to `template_options_controller`

`Rails.application.config.default_funder_id` is used multiple times in this controller, so defining a variable for it is best practice.

momo3404 committed:
Tag templates with appropriate source in controller

- Templates will need to be tagged appropriately to allow for organization into groups with headers in the template dropdown.
- The tag_templates function is used to add a source tag to each template.

momo3404 committed:
Refactor new.js to improve readability

- All possible error conditions are checked at the beginning to simplify conditional logic.
- `templates` is defined as a variable for `data.templates`
- `only` variable is renamed to `only_template`

momo3404 committed:
Add template headers through new.js

- Define arrays for ordering template groups
- Add appendGroup helper function to add templates to proper groups

momo3404 committed:
Add template header CSS

momo3404 committed:
Add "show more templates" button in new.js

- Add the `showMore` button if there are any remaining templates that can be displayed
- Add each template to a hiddenContainer that is displayed after the button is clicked

momo3404 committed:
Add "show more templates" CSS and styling improvements

momo3404 committed:
Break up `success()` function with helper functions

- Four helper functions were added to break up success()
- `categorizeTemplates()` - separates templates by their source
- `createTemplateItem()` - creates clickable template item
- `appendGroup()` - appends templates into appropriate groups
- `appendShowMoreSection()` - adds "Show more templates" button and hidden container to dropdown

aaronskiba commented on app/assets/stylesheets/blocks/_index.scss:
```diff
@@ -36,6 +36,7 @@
 @use 'spinner';
 @use 'statuses';
 @use 'tables';
+@use '../dmp-assistant/blocks/template_dropdown_menu';
```
I'm thinking this might not be necessary. You already have `@use 'template_dropdown_menu';` in `app/assets/stylesheets/dmp-assistant/blocks/_index.scss`, and that is being imported via `@import "dmp-assistant/blocks"` in `app/assets/stylesheets/application.scss`.

momo3404 replied:
You're right, I was following the process that was usesd with `password_checklist` in https://github.com/portagenetwork/roadmap/pull/1158 but I see it's redundant here.

aaronskiba commented:
It looks like these changes are breaking a test https://github.com/portagenetwork/roadmap/actions/runs/18754947740/job/53504332739?pr=1215.

This will be a heavily used feature, so it might be wise to also add a test to verify the headers and new dropdown behaviour. The new test could probably be added in `spec/features/plans_spec.rb`.

momo3404 committed:
Remove redundant stylesheet call

@use 'template_dropdown_menu'; already exists in app/assets/stylesheets/dmp-assistant/blocks/_index.scss

momo3404 committed:
Fix failing test and header typo

- The test was failing since it was using the old select behaviour. It has been fixed by making sure the test opens the dropdown and finds the correct CSS section.
- The header for organisations should be 'organiSational templates' not 'organisZational templates'

momo3404 committed:
Added test for dropdown headers and button

- Moved 'Rails.application.config.default_funder_id = @funding_org.id' such that @funding_org.id and the default_funder_id are the same in the test
- Added another funder template to test "show more templates"
- Tested for headers appearing correctly in dropdown

aaronskiba reviewed (commented):
I wonder if we should address the following:

After selecting an organisation, the select template dropdown is rendered. The "Please select a template" instruction is selected as the default option and I'm able to click "Create plan" with that default option. Doing so re-renders the "Create plan" page, as well as the flash message depicted in the second screenshot.

The error handling is doing what it should, but it might be even better to keep the keep the "Create plan" button disabled to start?

<img width="1223" height="388" alt="Screenshot from 2025-10-24 11-41-18" src="https://github.com/user-attachments/assets/c6eb571b-69b1-4a23-bbc1-549ee70c9d92" />
<img width="1223" height="603" alt="Screenshot from 2025-10-24 11-41-26" src="https://github.com/user-attachments/assets/22afadb6-783b-462b-a585-c044a3ebfb17" />

momo3404 commented:
> I wonder if we should address the following:
> 
> After selecting an organisation, the select template dropdown is rendered. The "Please select a template" instruction is selected as the default option and I'm able to click "Create plan" with that default option. Doing so re-renders the "Create plan" page, as well as the flash message depicted in the second screenshot.
> 
> The error handling is doing what it should, but it might be even better to keep the keep the "Create plan" button disabled to start?

Yes, I noticed that too and think it's worth addressing. The previous behaviour did make sure the "Create plan" button was disabled until a template was picked.

aaronskiba commented on app/javascript/src/plans/new.js:
```diff
@@ -21,32 +27,108 @@ $(() => {
-        error();
-      }
+    const templates = data.templates;
+    if (!isObject(data) || !isArray(templates) || templates.length === 0) return error();
```
I'm thinking it would be safer to assign `const templates = data.templates;` after the line that is currently just below it. That way there is no chance of `const templates = data.templates` throwing an error.

aaronskiba replied:
Or optional chaining (`data?.templates`) would probably work too.

aaronskiba commented on app/javascript/src/plans/new.js:
```diff
@@ -21,32 +27,108 @@ $(() => {
+    // If there is only one template, set the input field value and submit the form
+    // otherwise show the dropdown list and the 'Multiple templates found message'
+    if (templates.length === 1) {
+      const only_template = templates[0];
```
Probably better to just stick to camelcase throughout.

aaronskiba commented on app/javascript/src/plans/new.js:
```diff
@@ -21,32 +27,108 @@ $(() => {
-        error();
-      }
+    const templates = data.templates;
+    if (!isObject(data) || !isArray(templates) || templates.length === 0) return error();
```
Maybe it's okay, but prior to these changes, when `if (isObject(data) && isArray(data.templates))` was false, an `error()` was not returned.

aaronskiba commented on app/javascript/src/plans/new.js:
```diff
@@ -21,32 +27,108 @@ $(() => {
+    const title = $('<div>').text(template.title).html();
+    const item = $(`<a class="dropdown-item" href="#" data-id="${template.id}">${title}</a>`);
+
+    item.on('click', (e) => {
```
It might be best to add `e.preventDefault()` inside here. Otherwise, selecting a template would result in the browser jumping back to the top of the "Create plan" page.

momo3404 committed:
Edit `toggleSubmit()` to disable 'Create Plan' button correctly

- If no template is selected, tmplt has not value and thus this condition can be used to prevent the 'Create Plan' button from being clickable until a template has been selected.
- Defined submitButton to improve readability of the function.

momo3404 committed:
Edit new.js to follow better practices

- Add e.preventDefault() inside createTemplateItem. Without this, selecting a template might result in the browser jumping back to the top of the "Create plan" page.
-  Assign const templates = data.templates; such that there is no way it can throw an error.
- Rename only_template to follow camelCase

aaronskiba reviewed (commented):
I'm noticing some unexpected behaviour here:

Pic 1: I pick a template that belongs to the org that I selected (no problem here).

Pic 2: I check "no org", but the org specific template is still selected.

Pic 2: I pick another org, but the previous org's template is still selected.


<img width="1223" height="603" alt="Screenshot from 2025-10-27 09-38-25" src="https://github.com/user-attachments/assets/23937da4-c68c-4053-85ec-e65a9d74f676" />
<img width="1223" height="603" alt="Screenshot from 2025-10-27 09-38-31" src="https://github.com/user-attachments/assets/b60d62c8-b00c-4918-8bc7-e38b3bdf83e1" />

<img width="1223" height="603" alt="Screenshot from 2025-10-27 09-38-40" src="https://github.com/user-attachments/assets/1f5dbb78-2558-4398-8403-a68e9c9375c5" />

aaronskiba commented on app/assets/stylesheets/dmp-assistant/blocks/_template_dropdown_menu.scss:
```diff
@@ -0,0 +1,62 @@
+#template-dropdown-menu {
```
Overall, the styling is very nice. I wonder if we can re-use some of the existing branding variables here, or maybe the existing branding could even adopt some of these styling ideas?

I put up the following PR to help possibly faciliate that: https://github.com/portagenetwork/roadmap/pull/1218.

momo3404 replied:
We definitely could, I made a change [here](https://github.com/portagenetwork/roadmap/pull/1215/commits/a61d2e201639f47d5bb2702a5a3ab55d5e1a2793) to have the dropdown match the app's styling and use the existing branding variables.

momo3404 committed:
Add `resetTemplateDropdown()` to handle `no research org` and org change

- The function resetTemplateDropdown() resets the template dropdown's value and menu items.
- It will be used when the 'No research organisation' checkbox is toggled and when the org is changed in the 'Create Plan' page

momo3404 commented:
> I'm noticing some unexpected behaviour here:
> 
> Pic 1: I pick a template that belongs to the org that I selected (no problem here).
> 
> Pic 2: I check "no org", but the org specific template is still selected.
> 
> Pic 2: I pick another org, but the previous org's template is still selected.
> 

I added a fix [here](https://github.com/portagenetwork/roadmap/pull/1215/commits/913c6acd8f2cc871053ff04814f5787aeb003a28) that should address this. The issue was that the template dropdown was not resetting the value and menu items after the org was changed or the "No research organisation" checkbox was toggled.

aaronskiba commented on app/assets/stylesheets/dmp-assistant/blocks/_template_dropdown_menu.scss:
```diff
@@ -0,0 +1,62 @@
+#show-more-templates:focus {
+  outline: none;
+  box-shadow: none;
+}
```
It might better to use nesting here (also for `#template-dropdown-menu`):
```scss
#show-more-templates {
  width: 100%;
  background: transparent;
  border: none;
  text-align: center;
  font-family: inherit;
  font-weight: 500;
  padding: 0.5rem 0.75rem;
  cursor: pointer;

  &:hover {
    text-decoration: underline;
    background-color: rgba(0, 0, 0, 0.03);
  }

  &:focus {
    outline: none;
    box-shadow: none;
  }
}
```

momo3404 committed:
Edited dropdown styling to match app branding

- Changed the class of the "show more templates" button to match the default button styling in the app.
- Used alliance styling colors.
- Removed unnecessary styling specifications.

aaronskiba commented on app/javascript/src/plans/new.js:
```diff
@@ -21,34 +31,130 @@ $(() => {
+  };
+
+  // Reset dropdown when a new research organisation is selected
+  $('#research-org-controls input.autocomplete-result').on('change', resetTemplateDropdown);
```
The comment states "Reset dropdown when a new research organisation is selected", but I'm wondering if this listener is attached to the autocomplete input and is fired for every input change (e.g. adding or deleting a character).

If that is the case, it might be best to only trigger `resetTemlateDropdown` when a valid org is selected.

momo3404 committed:
Add nesting in dropdown CSS

momo3404 committed:
Update new.js to only call `resetTemplateDropdown()` when a valid org is selected

- The previous behaviour had the listener attached to the autocomplete input and was firing for every input change (e.g. adding or deleting a character).
- This new logic ensures the function is only called when a valid org is selected in the org dropdown or typed fully.

aaronskiba commented on app/javascript/src/plans/new.js:
```diff
@@ -21,33 +31,144 @@ $(() => {
+    const templates = data.templates;
+
+    // Categorize templates based on their source tag
+    const { orgTemplates, priorityTemplates, otherTemplates } = categorizeTemplates(templates);
```
Do you think it would be better to do this within TemplateOptionsController?

momo3404 replied:
I think TemplateOptionsController already handles this with the tagging. All this does is just check the tagging from the controller and append the templates to the appropriate array.

aaronskiba commented on app/javascript/src/plans/new.js:
```diff
@@ -5,14 +5,24 @@ import { isObject, isArray, isString } from '../utils/isType';
-    if (isString(tmplt)) {
-      $('#new_plan button[type="submit"]').removeAttr('disabled')
-        .removeAttr('data-toggle').removeAttr('title');
+  const toggleSubmit = (select_dropdown = true) => {
```
Just a little thing with the camel case again.

momo3404 committed:
Fix camelcase naming

aaronskiba committed:
Simplify template handling / improve payload structure

Simplify template handling by returning a structured JSON payload:

- Backend now returns:
  {
    templates: {
      org_templates: [...],
      priority_templates: [...],
      other_templates: [...]
    },
    total_templates: <count>
  }

- Eliminates the need to tag each template individually with a source.
- Frontend no longer needs to categorize templates manually.
- Minor frontend adjustments made to consume the new payload structure.

aaronskiba committed:
Simplify `spec/features/plans_spec.rb`

These changes address an unnecessary number of templates being created inside the `before do` block. Some redundancies have been removed from the creation of `@extra_funder_template` as well.

momo3404 committed:
Fix dropdown extending beyond footer

- The previous hardcoded max-height was not working properly, as sometimes the dropdown's max height would extend past the footer.
- Helper function adjustDropdownMaxHeight() was added to dynamically assign the max-height to always be the middle of the footer.
- Function is called in success() after template dropdown becomes visible (as helper cannot work without the values from the DOM)

aaronskiba commented on app/javascript/src/plans/new.js:
```diff
@@ -5,14 +5,24 @@ import { isObject, isArray, isString } from '../utils/isType';
 import { renderAlert, hideNotifications } from '../utils/notificationHelper';
 
 $(() => {
-  const toggleSubmit = () => {
```
To reduce repeated DOM queries and improve readability, I would recommend caching repeatedly used selectors. For example, assign `const $planTemplateId = $('#plan_template_id')` at the top of the main callback, and reference that.

momo3404 replied:
Yes, I did something similar with `const submitButton = $('#new_plan button[type="submit"]');` and that can definitely be added for other selectors.

aaronskiba commented on app/javascript/src/plans/new.js:
This file is getting large and will be hard to maintain over time. It might be worth refactoring into smaller modules. For example, separating dropdown rendering, autocomplete handling, and AJAX logic into their own files. I could give that a shot if you'd like.

momo3404 replied:
I was thinking that too, I can definitely work on that for now.

aaronskiba commented on app/javascript/src/plans/new.js:
```diff
@@ -1,18 +1,28 @@
-        .attr('data-toggle', 'tooltip').attr('title', getConstant('NEW_PLAN_DISABLED_TOOLTIP'));
+      tmplt = $('#plan_template_id').val();
+    }
+
```
Would the following simplification work here?

```js
$(() => {
  const toggleSubmit = () => {
    const tmplt = $('#plan_template_id').val();

    const submitButton = $('#new_plan button[type="submit"]');
```

momo3404 replied:
`$('#plan_template_id').find(':selected').val()` is intended for use with `select_tag`, so I didn't want to remove that functionality in case another existing `select_tag` was also using this function.

aaronskiba replied:
I'm thinking `app/javascript/src/plans/new.js` is only run on the `plans/new` page. Since our changes included replacing the `select_tag` with the `hidden_field_tag`, I think we can make the simplification here.

aaronskiba committed:
Cleanup `plans/new.js` after switching to dropdown

- Removed select_tag-specific code from toggleSubmit:
  - No longer checks `:selected` since `plan_template_id` is now a hidden field.
- Removed dead lines related to the old select_tag:
  - `$('#plan_template_id').find(':selected').removeAttr('selected')`
  - `$('#plan_template_id option').remove()`
- Simplifies and clarifies template selection handling with the new dropdown implementation.

aaronskiba committed:
Remove redundant org change listener and consolidate template reset

- Removed `#research-org-controls input.autocomplete-result` change handler:
  - Its functionality is already handled by `handleComboboxChange`.
  - Avoids unnecessary JSON parsing and constant firing on every keystroke.

- Added `resetTemplateDropdown()` call inside `handleComboboxChange`:
  - Simplifies the function by removing explicit lines and instead handling them conditionally inside `resetTemplateDropdown()`
  - Consolidates template dropdown reset logic in a single place, reducing redundant event handlers.

momo3404 committed:
Cache repeatedly used selectors

Cache redundant DOM queries with variables

(In adjustDropdownMaxHeight) Replace redundant DOM queries with unwrapping cached jQuery objects using `.get(0)` (https://api.jquery.com/get/).

aaronskiba committed:
Optimise click handling of template items

Instead of binding click handlers to each individual template item, this change creates a single delegated handler on `templateDropdownMenu`.
- Listens at the container level; clicks on child elements bubble up to the handler.
- Adds `data-title` attribute to template items. (Allows us to still set `templateDropdown.text(template.title)`)
- Improves performance and reduces memory usage.

aaronskiba committed:
Refactor dropdown sizing into adjustDropdownMaxHeight

- Consolidate all template dropdown height logic into `adjustDropdownMaxHeight()`.
- Simplifies `success()` by removing inline dropdown sizing logic.

aaronskiba committed:
Use jQuery element API in `createTemplateItem`

- Stop using raw HTML templates for dropdown items.
- Use jQuery’s attribute map and `.text()` to properly escape content and keep `data-*` attributes on the element.

aaronskiba committed:
Enable translations for `templateDropdown` text

Refactored templateDropdown strings to support in-app translation.
- These values must be defined and translated on the Rails side first.

aaronskiba committed:
Add width resizing & refactor templateDropdownMenu

- Renamed `adjustDropdownMaxHeight()` to `resizeTemplateDropdownMenu()`
- Set templateDropdownMenu.outerWidth(templateDropdown.outerWidth()) to match dropdown button width, preventing resize when "Show more..." is clicked
- Simplified the `isVisible` check
- Updated template dropdown button ID and references from `templateDropdown` to `template-dropdown` for consistency with other selectors

aaronskiba committed:
Fix flash of template section on `/plans/new` load

`#available-templates` was momentarily visible before JavaScript hid it.
Set it hidden in the markup (`display: none`) to ensure it remains invisible until shown intentionally. This eliminates the visual flash during page load.

aaronskiba committed:
Add defences to template response handling

- Added isValidTemplateResponse helper to validate the structure and content of AJAX template responses.
- Checks now ensure `data.total_templates` is a number greater than 0 and that each template group (`org_templates`, `priority_templates`, `other_templates`) is an array.
- Added a guard to ensure onlyTemplate is defined before accessing its properties.
- Improves safety against malformed or unexpected server responses.

aaronskiba committed:
Refactor template data extraction into helper

- Add extractTemplateData() helper to handle destructuring of total_templates and template groups (org_templates, priority_templates, other_templates).
- Now, rather than dealing with the template response structure, success() works with clearly defined variables returned from extractTemplateData().

aaronskiba commented on app/assets/stylesheets/dmp-assistant/blocks/_template_dropdown_menu.scss:
```diff
@@ -0,0 +1,48 @@
+    padding: 0.5rem 0.75rem;
+    cursor: pointer;
+  }
+}
```
This is just a comment / possible future to-do:

I prefer this styling over what currently exists for the other dropdowns. It might be best to eventually generalise some of these ideas to all of the other dropdowns, dropdown items, etc.

aaronskiba commented on app/javascript/src/plans/new.js:
```diff
@@ -21,31 +33,142 @@ $(() => {
+  // Helper for resizing width and height of templateDropdownMenu
+  // Matches width to that of the dropdown button (prevents resizing when "Show more..." is clicked)
+  // Sets max-height to avoid extending beyond the footer
+  const resizeTemplateDropdownMenu = () => {
+    const templateDropdownEl = templateDropdown.get(0);
+    const templateDropdownMenuEl = templateDropdownMenu.get(0);
+    const footerEl = document.getElementById('footer-navbar');
+
+    // Return early if any required element is missing or dropdown is not visible
+    if (!(templateDropdownEl && templateDropdownMenuEl && footerEl  && templateDropdown.is(':visible'))) return;
+
+    // Set width
+    templateDropdownMenu.outerWidth(templateDropdown.outerWidth());
+
+    // Set maxHeight
+    const dropdownRect = templateDropdownEl.getBoundingClientRect();
+    const footerRect = footerEl.getBoundingClientRect();
+
+    // Ideal location for dropdown to end is middle of the footer
+    const footerMiddle = (footerRect.bottom + footerRect.top) / 2;
+
+    const spaceAvailable = footerMiddle - dropdownRect.bottom;
+
+    if (spaceAvailable > 0) {
+      templateDropdownMenuEl.style.maxHeight = `${spaceAvailable}px`;
+    }
+  };
```
This is a good addition. Other dropdowns could also benefit from this extra handling. In the future, we could bind the resizing to window.resize so it adjusts automatically.

aaronskiba committed:
Update template validation on plan creation

Restrict templates that can be associated with newly created plans. Only templates that are published and unarchived are now allowed.

aaronskiba committed:
Update CHANGELOG.md

aaronskiba committed:
Merge pull request #1225 from portagenetwork/aaron/fix/template-validation-on-plan-creation

Update template validation on plan creation

momo3404 committed:
Add changelog entry
